File: data-structures/lists/java/nested_lists_add.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

import java.util.ArrayList;
import java.util.Scanner;

class NestedLists  {
    // The main method is the entry point for all Java programs
    public static void main(String[] args) {
        // Declare and initialise the nested list of astronauts
        ArrayList<ArrayList<String>> allAstronauts = new ArrayList<ArrayList<String>>();
        initialiseAstronauts(allAstronauts);

        // Use Scanner for reading user input from the console
        Scanner scanner = new Scanner(System.in);

        // Ask the user to input the new astronaut's details
        System.out.println("Enter the spacecraft's name:");
        String spacecraft = scanner.nextLine();

        System.out.println("Enter the astronaut's full name:");
        String name = scanner.nextLine();

        System.out.println("Enter the current number of days in space:");
        String days = scanner.nextLine();

        // Add the new astronaut to the list
        addNewAstronaut(allAstronauts, spacecraft, name, days);

        // Output the details of the new astronaut
        int lastIndex = allAstronauts.size() - 1;
        ArrayList<String> astronaut = allAstronauts.get(lastIndex);
        displayAstronautDetails(astronaut);
    }

    // Add a new astronaut list to the end of the nested list
    public static void addNewAstronaut(ArrayList<ArrayList<String>> allAstronauts,
                                       String givenSpacecraft, String givenName, String givenDays) {

        // Create a new list for the astronaut
        ArrayList<String> newAstronaut = new ArrayList<String>();
        newAstronaut.add(givenSpacecraft);
        newAstronaut.add(givenName);
        newAstronaut.add(givenDays);

        // Add the new astronaut list to the end of the nested list
        allAstronauts.add(newAstronaut);
    }

    // Display the values of each astronaut in the given list
    public static void displayAllAstronauts(ArrayList<ArrayList<String>> allAstronauts) {
        // Repeat for each astronaut in the outer list
        for (ArrayList<String> astronaut : allAstronauts) {
            // Repeat for each value in the inner list
            for (String value : astronaut) {
                System.out.print(value + "  "); // Do not print new line
            }
            System.out.println(); // Print a new line after each astronaut
        }
    }

    // Initialise the nested list of astronauts
    public static void initialiseAstronauts(ArrayList<ArrayList<String>> allAstronauts) {
        ArrayList<String> first = new ArrayList<String>();
        first.add("ISS");
        first.add("Sergey Prokopyev");
        first.add("176");
        allAstronauts.add(first);

        ArrayList<String> second = new ArrayList<String>();
        second.add("ISS");
        second.add("Dmitry Petelin");
        second.add("176");
        allAstronauts.add(second);

        ArrayList<String> third = new ArrayList<String>();
        third.add("ISS");
        third.add("Frank Rubio");
        third.add("176");
        allAstronauts.add(third);

        ArrayList<String> fourth = new ArrayList<String>();
        fourth.add("Tiangong");
        fourth.add("Fei Junlong");
        fourth.add("112");
        allAstronauts.add(fourth);

        ArrayList<String> fifth = new ArrayList<String>();
        fifth.add("Tiangong");
        fifth.add("Deng Qingming");
        fifth.add("107");
        allAstronauts.add(fifth);

        ArrayList<String> sixth = new ArrayList<String>();
        sixth.add("Tiangong");
        sixth.add("Zhang Lu");
        sixth.add("107");
        allAstronauts.add(sixth);

        ArrayList<String> seventh = new ArrayList<String>();
        seventh.add("ISS");
        seventh.add("Stephen Bowen");
        seventh.add("54");
        allAstronauts.add(seventh);

        ArrayList<String> eighth = new ArrayList<String>();
        eighth.add("ISS");
        eighth.add("Warren Hoburg");
        eighth.add("14");
        allAstronauts.add(eighth);

        ArrayList<String> ninth = new ArrayList<String>();
        ninth.add("ISS");
        ninth.add("Sultan Al Neyadi");
        ninth.add("14");
        allAstronauts.add(ninth);

        ArrayList<String> tenth = new ArrayList<String>();
        tenth.add("ISS");
        tenth.add("Andrey Fedyaev");
        tenth.add("14");
        allAstronauts.add(tenth);
    }
}

File: data-structures/lists/python/nested_lists_add.py
# Raspberry Pi Foundation
# Developed as part of Ada Computer Science
# 
# Usage licensed under CC BY-NC-SA 4.0


def add_new_astronaut(all_astronauts, given_spacecraft, given_name, given_days):
    """Add a new astronaut list to the end of the nested list"""

    # Create a new list for the astronaut
    new_astronaut = [given_spacecraft, given_name, given_days]

    # Add the new astronaut list to the end of the nested list
    all_astronauts.append(new_astronaut)


def display_astronaut_details(astronaut):
    """Display the values from the given astronaut list"""

    print("\nAstronaut details:")
    
    # Repeat for each value in the list
    for value in astronaut:
        print(value)


def main():
    # Declare and initialise the nested list of astronauts
    all_astronauts = [
        ["ISS", "Sergey Prokopyev", "176"],
        ["ISS", "Dmitry Petelin", "176"],
        ["ISS", "Frank Rubio", "176"],
        ["Tiangong", "Fei Junlong", "112"],
        ["Tiangong", "Deng Qingming", "107"],
        ["Tiangong", "Zhang Lu", "107"],
        ["ISS", "Stephen Bowen", "54"],
        ["ISS", "Warren Hoburg", "14"],
        ["ISS", "Sultan Al Neyadi", "14"],
        ["ISS", "Andrey Fedyaev", "14"]
    ]

    # Ask the user to input the new astronaut's details
    spacecraft = input("Enter the name of the spacecraft: ")
    name = input("Enter the astronaut's full name: ")
    days = input("Enter the current number of days in space: ")

    # Add the new astronaut to the list
    add_new_astronaut(all_astronauts, spacecraft, name, days)

    # Output the details of the new astronaut
    last_index = len(all_astronauts) - 1
    astronaut = all_astronauts[last_index]
    display_astronaut_details(astronaut)


# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == "__main__":
    main()
